File: packages/sdk/hush-py/src/clawdstrike/hunt/stream.py
"""Async streaming for Hunt SDK -- yields events and alerts from NATS."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import AsyncIterator, Union

from clawdstrike.hunt.correlate import CorrelationEngine
from clawdstrike.hunt.errors import WatchError
from clawdstrike.hunt.timeline import parse_envelope
from clawdstrike.hunt.types import Alert, TimelineEvent, WatchConfig

_NATS_SUBJECT = "clawdstrike.sdr.fact.>"


@dataclass(frozen=True)
class StreamAlertItem:
    """A stream item wrapping an alert."""

    type: str
    alert: Alert


@dataclass(frozen=True)
class StreamEventItem:
    """A stream item wrapping a timeline event."""

    type: str
    event: TimelineEvent


StreamItem = Union[StreamAlertItem, StreamEventItem]


async def stream(config: WatchConfig) -> AsyncIterator[Alert]:
    """Stream alerts from NATS as an async iterator.

    Subscribes to ``clawdstrike.sdr.fact.>``, parses spine envelopes into
    timeline events, feeds them through a :class:`CorrelationEngine`, and
    yields alerts when correlation rules fire.

    Args:
        config: Watch configuration (NATS URL, rules, etc.).

    Yields:
        Alerts from the correlation engine.

    Raises:
        WatchError: If the ``nats-py`` package is not installed.
    """
    try:
        import nats as nats_mod  # type: ignore[import-untyped]
    except ImportError as exc:
        raise WatchError(
            "The 'nats-py' package is required for streaming. "
            "Install it with: pip install nats-py"
        ) from exc

    engine = CorrelationEngine(list(config.rules))
    nc = await nats_mod.connect(
        config.nats_url,
        **({"user_credentials": config.nats_creds} if config.nats_creds else {}),
    )
    sub = await nc.subscribe(_NATS_SUBJECT)

    try:
        async for msg in sub.messages:
            try:
                envelope = json.loads(msg.data.decode())
            except (json.JSONDecodeError, UnicodeDecodeError):
                continue
            if not isinstance(envelope, dict):
                continue

            event = parse_envelope(envelope)
            if event is None:
                continue

            alerts = engine.process_event(event, config.max_window)
            for alert in alerts:
                yield alert
    finally:
        await nc.drain()


async def stream_all(config: WatchConfig) -> AsyncIterator[StreamItem]:
    """Stream all items (events and alerts) from NATS as an async iterator.

    Like :func:`stream`, but yields both events and alerts as tagged
    :class:`StreamItem` objects.

    Args:
        config: Watch configuration (NATS URL, rules, etc.).

    Yields:
        :class:`StreamEventItem` for each parsed event and
        :class:`StreamAlertItem` for each fired alert.

    Raises:
        WatchError: If the ``nats-py`` package is not installed.
    """
    try:
        import nats as nats_mod  # type: ignore[import-untyped]
    except ImportError as exc:
        raise WatchError(
            "The 'nats-py' package is required for streaming. "
            "Install it with: pip install nats-py"
        ) from exc

    engine = CorrelationEngine(list(config.rules))
    nc = await nats_mod.connect(
        config.nats_url,
        **({"user_credentials": config.nats_creds} if config.nats_creds else {}),
    )
    sub = await nc.subscribe(_NATS_SUBJECT)

    try:
        async for msg in sub.messages:
            try:
                envelope = json.loads(msg.data.decode())
            except (json.JSONDecodeError, UnicodeDecodeError):
                continue
            if not isinstance(envelope, dict):
                continue

            event = parse_envelope(envelope)
            if event is None:
                continue

            yield StreamEventItem(type="event", event=event)

            alerts = engine.process_event(event, config.max_window)
            for alert in alerts:
                yield StreamAlertItem(type="alert", alert=alert)
    finally:
        await nc.drain()


__all__ = [
    "stream",
    "stream_all",
    "StreamItem",
    "StreamAlertItem",
    "StreamEventItem",
]

File: packages/sdk/clawdstrike-hunt/src/stream.ts
/**
 * Async generator streaming — subscribe to NATS for live envelope events,
 * feed them through the CorrelationEngine, and yield alerts/events.
 *
 * Requires the optional `nats` peer dependency.
 */

import { parseEnvelope } from './timeline.js';
import { CorrelationEngine } from './correlate/engine.js';
import { WatchError } from './errors.js';
import { buildNatsConnectOptions } from './nats.js';
import type { Alert, TimelineEvent, WatchConfig } from './types.js';

const NATS_SUBJECT = 'clawdstrike.sdr.fact.>';

export interface StreamOptions extends WatchConfig {
  signal?: AbortSignal;
}

export type StreamItem =
  | { type: 'alert'; alert: Alert }
  | { type: 'event'; event: TimelineEvent };

/**
 * Stream alerts as an async iterable. Yields only alerts.
 */
export async function* stream(
  options: StreamOptions,
): AsyncGenerator<Alert, void, undefined> {
  const natsModuleName = 'nats';
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let natsModule: any;
  try {
    natsModule = await import(natsModuleName);
  } catch {
    throw new WatchError(
      "The 'nats' package is required for streaming. Install it with: npm install nats",
    );
  }

  const engine = new CorrelationEngine(options.rules);
  const nc = await natsModule.connect(
    await buildNatsConnectOptions(natsModule, options.natsUrl, options.natsCreds),
  );
  const sub = nc.subscribe(NATS_SUBJECT);

  if (options.signal) {
    const onAbort = () => sub.unsubscribe();
    options.signal.addEventListener('abort', onAbort, { once: true });
  }

  try {
    for await (const msg of sub) {
      if (options.signal?.aborted) break;

      let envelope: unknown;
      try {
        envelope = JSON.parse(new TextDecoder().decode(msg.data));
      } catch {
        continue;
      }

      const event = parseEnvelope(envelope);
      if (!event) continue;

      const alerts = engine.processEvent(event, options.maxWindow);
      for (const alert of alerts) {
        yield alert;
      }
    }

    // Flush remaining partial windows on shutdown.
    const remaining = engine.flush();
    for (const alert of remaining) {
      yield alert;
    }
  } finally {
    await nc.drain();
  }
}

/**
 * Stream all items (events and alerts) as an async iterable.
 */
export async function* streamAll(
  options: StreamOptions,
): AsyncGenerator<StreamItem, void, undefined> {
  const natsModuleName = 'nats';
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let natsModule: any;
  try {
    natsModule = await import(natsModuleName);
  } catch {
    throw new WatchError(
      "The 'nats' package is required for streaming. Install it with: npm install nats",
    );
  }

  const engine = new CorrelationEngine(options.rules);
  const nc = await natsModule.connect(
    await buildNatsConnectOptions(natsModule, options.natsUrl, options.natsCreds),
  );
  const sub = nc.subscribe(NATS_SUBJECT);

  if (options.signal) {
    const onAbort = () => sub.unsubscribe();
    options.signal.addEventListener('abort', onAbort, { once: true });
  }

  try {
    for await (const msg of sub) {
      if (options.signal?.aborted) break;

      let envelope: unknown;
      try {
        envelope = JSON.parse(new TextDecoder().decode(msg.data));
      } catch {
        continue;
      }

      const event = parseEnvelope(envelope);
      if (!event) continue;

      yield { type: 'event' as const, event };

      const alerts = engine.processEvent(event, options.maxWindow);
      for (const alert of alerts) {
        yield { type: 'alert' as const, alert };
      }
    }

    const remaining = engine.flush();
    for (const alert of remaining) {
      yield { type: 'alert' as const, alert };
    }
  } finally {
    await nc.drain();
  }
}
